### ISSUE TITLE
Precision 1 and recall 0.000589

### ISSUE THREAD

**Issue by madr3z** [WongKinYiu/ScaledYOLOv4#255]
I keep getting a precision of 1 and a recall of 0.000589 from the 3rd epoch all the way to the 300th. I'm training binary detection model and have used `--single-cls`. I do get precision < 1, and recall >  0.000589, when training the same validation and training set on other libraries (eg. Darknet Yolov4, Ultralytics Yolov5 etc.). So it seems that the problem might be with the metrics calculation in ScaledYOLOv4.
I have looked around the code and couldn't find the problem. What do you think it might be?

**Comment by WongKinYiu** [comment]
which branch you are using?
does the loss shows nan?

**Comment by madr3z** [comment]
I am using Yolov4-CSP branch. Loss progress looks fine and no nan values. mAP progress looks fine and no nan values.

**Comment by WongKinYiu** [comment]
if mAP is low, you could try this hyp.
```
lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.2  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.5  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 1.0  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
# anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
hsv_h: 0.015  # image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # image HSV-Value augmentation (fraction)
degrees: 0.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.5  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # image flip up-down (probability)
fliplr: 0.5  # image flip left-right (probability)
mosaic: 1.0  # image mosaic (probability)
mixup: 0.0  # image mixup (probability)
```

**Comment by madr3z** [comment]
mAP is fine. Actually impressive (94%)! I wanted to know the precision and recall since the model is performing well but it seems that something is wrong with their calculations.

**Comment by WongKinYiu** [comment]
you could modify the utils/metrics.py according to Ultralytics Yolov5.
i use the calculation in release v3.0, and the metric has been modified in newer release version.

**Comment by madr3z** [comment]
Thank you for your advice.

**Comment by liuwenhua6666** [comment]
lr0 设置小点
